docs/proposals/pod-troubleshooting.md

```diff
@@ -0,0 +1,489 @@
+
+The disadvantage of using containers built `FROM scratch` is the lack of system
+binaries provided a base Operation System image makes it difficult to
+troubleshoot running containers. Kubernetes should enable troubleshooting pods
+regardless of the contents of the container image.
+
+## Goals and Non-Goals
+
+Goals include:
+
+*   Enable troubleshooting of minimal container images
+*   Allow troubleshooting of containers in `CrashLoopBackoff`
+*   Improve supportability of native Kubernetes applications
+*   Enable novel uses of short-lived containers in pods
+
+Non-Goals of this proposal are:
+
+*   Fully design the ability to cross-mount container image filesystems. This is
+    complex and should be covered in a separate doc.
+*   Guarantee resources for ad-hoc troubleshooting. If troubleshooting causes a
+    pod to exceed its resource limit it may be evicted.
+
+## Requirements
+
+A solution to troubleshoot minimal container images MUST:
+
+*   troubleshoot arbitrary running containers with minimal prior configuration
+*   allow access to all namespaces shared by containers in a pod and the mount
+    namespaces of individual containers
+*   fetch troubleshooting utilities at debug time rather than at the time of pod
+    initialization
+*   respect admission restrictions
+*   allow introspection of pod state using existing tools (no hidden containers)
+*   support arbitrary runtimes via the CRI (possibly with reduced feature set)
+
+A good solution SHOULD:
+
+*   have an excellent user experience (i.e. should be a feature of the platform
+    rather than a config-time solution)
+*   require no direct access to the node
+*   have no inherent side effects to the running container image
+*   re-use existing container image distribution channels
+*   enable detection of pods that have been modified by an exec ("tainted" pods)
+
+## kubectl debug
+
+We will introduce an "execute a debug container in a pod" pattern that parallels
+`kubectl exec`'s "execute a debug process in a container" in a new command,
+`kubectl debug`, which has two modes of operation:
+
+1.  *Copy Debug* mode creates a copy of a pod with minor changes to the pod spec
+    specified on the command line.
+1.  *Running Debug* mode causes Kubernetes to run a new *debug container* in the
+    pod context. A *debug container* is not part of the pod spec and cannot be
+    configured by the user. It is created by a debug operation on an existing
+    pod and reported in `PodStatus`.
+
+The mode distinction is for the purposes of this proposal, and the end user need
+not be aware of them. They form a single debugging command that allows the user
+to modify a running pod if possible and create a copy otherwise. The
+determination is made based on container name, for example:
+
+```
+% kubectl run example --image=nginx --restart=Never
+pod "example" created
+
+# Container "shell" does not exist, add a new container to running pod
+kubectl debug -p example -c shell --image=debian
```

Comment:
I would like to propose to change the syntax to `kubectl debug POD [args …]` where `POD` is *always* the pod you'd like to debug. The current proposal has the confusing semantics that the pod you want to debug is sometimes passed as `-p POD` and sometimes as `--copy-of POD`.

I'll try to attempt to propose an alternative syntax by example:

```
# Container "shell" does not exist, add a new container to running pod
# N.B.: --in-place is borrowed from GNU's sed.
kubectl debug example --in-place -c shell --image=debian 

# Container DNE, create a copy with an additional container
# If -p/--pod is not given, a name is automatically derived.
kubectl debug example [-p example-copy] -c shell --image=debian

# Container name exists, create a copy with a different entrypoint 
kubectl debug example [-p example-copy] -c container-with-distro --command -- /bin/sh

# ERROR: cannot modify entrypoint of running pod (pod spec is immutable)
# (aka. --in-place and --command are mutually exclusive)
kubectl debug example --in-place -c container-with-distro --command -- /bin/sh # ERROR! 
```

I'd argue this has the following benefits:

*   `kubectl debug POD-TO-DEBUG` is consistent (both, across the different modes and other kubectl subcommands), easy to memorize and intuitive.
*   Inverting the logic of `--copy` to `--in-place` allows us to implement the *copy debug mode* first and add the *running debug mode* later without clumsy defaults or break the CLI.

    With `--copy`, we'd either
    *   have to require people type `--copy` always (even though the flag wouldn't serve a purpose at first) and make *running debug mode* the default when it's available,
    *   end up with `--nocopy` which is not very descriptive because is disables implicit behavior, or
    *  change the default behavior once *running debug mode* is implemented.

I'm not particularly happy with specifying the name of the cloned pod via `-p CLONED-POD` but I fail to come up with a good alternative. Thoughts?